[CS] Simplify `Solution::resolveInterfaceType`

hamishknight opened the pull request:
Rather than attempting to re-implement `simplifyType`, tweak `Solution::simplifyType` such that it can map the resulting type out of context, and can turn type variables into their opened generic parameters.

Resolves #77924

hamishknight commented:
@swift-ci please test

xedin reviewed (approved):
Thanks!

hamishknight commented:
Actually wait, this isn't quite right, we ought to still be recursing into types that have archetypes in case there are any type variables that need resolving...

xedin commented:
Oh, I didn't realize that we can have type variables inside of archetypes...

hamishknight commented:
I mean something like `($T0, <archetype>)`, `hasArchetype` will be `true`, but we still want to recurse to resolve the type variable. Really `hasArchetype` should be `is<ArchetypeType>`

hamishknight commented:
Updated, also noticed we can remove a recursive call

hamishknight commented:
@swift-ci please test

xedin commented:
Sure, makes sense!

xedin commented:
@hamishknight Looking at the whole thing, I wonder if it could be: `simplifyType(type)->mapTypeOutOfContext()` because dependent member type handling is what `simplifyType` does as well.

hamishknight commented:
IIRC the main motivation behind doing the transform manually was this:

```cxx
      // If this type variable is for a generic parameter, return that.
      if (auto *gp = tvt->getImpl().getGenericParameter())
        return gp;
```

i.e if we have `(T) -> Void` where `T == Int`, we want to return `(T) -> Void`, not `(Int) -> Void`. That being said, I don't know if this is really still worth it given we use `getInterfaceType` when we have a callee anyway.

The only interesting regression I can immediately think of is this:

```swift
func foo<T>(_ x: T) {}

func bar(_ fn: () -> Void) {
  (a: 0, b: foo).b(fn) // error: converting non-escaping parameter 'fn' to generic parameter 'T' may allow it to escape
}
```

This would become:

```
error: passing non-escaping parameter 'fn' to function expecting an @escaping closure
```

Maybe it's not worth it though, especially as `resolveInterfaceType` is already missing handling for pack types

xedin commented:
Yeah, I think the diagnostic change is okay here. Trying to re-invent for `simplifyType` does usually resolves in more problems in the long run, so we ought to use a flag or something to mimic this generic parameter behavior…

xedin commented:
Maybe we can even prevent opening during solving for cases like this?

hamishknight commented:
Updated to just call into `simplifyType` 👍

xedin commented:
Thank you! I think it might be worth opening an issue about opening happening in some places where it shouldn’t.

hamishknight commented:
Hmm, I think the opening behavior is correct though? i.e this is fine:

```swift

func foo<T>(_ x: T) {}

func bar() {
  (a: 0, b: foo).b(0)
}
```

xedin commented:
Your comment made it seem like we are opening in places where we shouldn't but I think I understand what you meant now. One more suggestion to make it work the same way - instead of `simplifyType` we can use `simplifyTypeImpl` in `resolveInterfaceType` (`Solution` is already a friend of `ConstraintSystem`) and provide a custom `getFixedTypeFn` that would produce underlying generic parameter type.

hamishknight commented:
Yeah I was thinking of something similar, except as a parameter to `simplifyType`, how does this look?

hamishknight commented:
@swift-ci please test

xedin commented:
Do you think the flag would be useful for anywhere but `resolveInterfaceType`, might be an overkill to leave both maybe?

hamishknight commented:
You mean remove `resolveInterfaceType` in favor of `simplifyType`? Don't mind doing that

xedin commented:
Yeah, if we want to go with the flag I’d think resolve method should go otherwise make the logic part of the resolve method…

hamishknight committed:
[CS] Simplify `Solution::resolveInterfaceType`

Rather than attempting to re-implement `simplifyType`,
tweak `Solution::simplifyType` such that it can
map the resulting type out of context, and can
turn type variables into their opened generic
parameters.

hamishknight committed:
[CS] Remove `resolveInterfaceType`

Use `simplifyType` instead with the new parameter
for getting an interface type. Also avoid using
`resolveInterfaceType` in CSApply since we don't
need the opened generic parameter mapping behavior.

hamishknight commented:
Done :)

hamishknight commented:
@swift-ci please test

xedin commented:
Great, thank you!
